text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+#### Discussion
+
+An initial implementation of this proposal would almost certainly provide only the soft ordering, since it practically a waypoint to get to the hard ordering implementation.
+
+Initial Validation are made on-chain prior to the Availabilty Protocol begins for any given Work Package. This ensures that the Work Package is still in scope $-$ i.e. recent enough and on the proper fork. However, this does not ensure that the Work Package is still in scope at the point that the Work Results are actually Accumulated. It is as yet unclear whether this is especially problematic.
+
+The same scope limit could be placed also on Accumulation; in neither variant does it introduce much additional complexity. In both cases it would require the same course of action as of Availability timing out without the possibility of retry. However whereas in the soft-variant we would not expect to see very different dynamics since one such time-out has no repurcissions beyond preventing the Accumulation of those Work Results, in the hard-ordering variant, it could mean a substantially greater occurance of the cascading failure logic calling into question the real purpose of the scoping: is it to protect the Validators from having to deal with indefinitely valid Work Packages or is it to protect the Accumulation logic from having to deal with the Results of older Work Packages?
+
+### Relay-chain Storage Pallet
+
+There is a general need to be able to reference large, immutable and long-term data payloads both on-chain and in-core. This is both the case for fixed-function logic such as fetching the VM code for `refine` and `accumulate` as well as from within Work Packages themselves.
+
+Owing to the potential for forks and disputes to happen beyond the scope of initial validation, there are certain quite subtle requirements over what data held on-chain may be utilized in-core. Because of this, it makes sense to have a general solution which is known to be safe to use in all circumstances. We call this solution the *Storage Pallet*.
+
+The Storage Pallet provides a simple API, accessible to untrusted code through host functions & extrinsics and to trusted Relay-chain code via a trait interface.
+
+```rust
+trait Storage {
+    /// Immutable function to attempt to determine the preimage for the given `hash`.
+    fn lookup(hash: &[u8; 32]) -> Option<Vec<u8>>;
```

Comment:
The main issue with `lookup` (when called in-core) is that results may differ based on the actual underlying state. Work package execution (in-core) needs to be return the same result in any unfinalized relay-chain state. So this pallet must be designed such that the result of `lookup` is always consistent for the same `Context` even in different relay-chain states.

Otherwise:
1. Data could be provided between a validator attesting to a Work Package and it being reported and available, leading to the former getting `None` and the latter getting `Some`.
2. Data could be unrequested between a validator attesting to a Work Package and it being reported and available, leading to the former getting `Some` and the latter getting `None`.

It looks like `lookup` needs to take some information about the desired `Context` (a relay chain block number is likely sufficient) in order to selectively answer with `None` even when the underlying preimage is known to the pallet.

Reply:
> The main challenge with lookup (when called in-core) is that results may differ based on the actual underlying state. Work package execution (in-core) needs to return the same result in any relay chain state in the span of possible states between initial attestation and disputes. So this pallet must be designed such that the result of lookup is always consistent for the same Context even in different relay-chain states.

Yes, this is exactly the point of having a specialised pallet for it.

Reply:
> It looks like lookup needs to take some information about the desired Context (a relay chain block number is likely sufficient) 

Why would this ever be different to the block number specified in the WP's `context`?

Reply:
Either:
  1. The relay chain validators all call `Storage::lookup` against the same relay chain state root (the context's), even up to 24 hours later.
  2. The relay chain validators can call `Storage::lookup` against different relay chain state roots but get exactly the same results based on the `context` block number.

The interface here presumes (1), which requires relay chain states not to be pruned for 24 hours.

If we do not want to require relay chain states for 24 hours (we currently don't - pruning is quite eager) then the interface must be adjusted as described above to prevent consensus splits.

Reply:
Lookup should provide the preimage. The preimage must be actively referenced within the Storage pallet at the `context` block for the lookup to succeed. The Storage pallet guarantees that this preimage remains on-chain (for the case of disputes) for 24 hours after all references are removed. I see no reason why any additional block would need to be specified.

Reply:
As I mentioned, there are race conditions. Sometimes the blob is available on-chain and needs to look like it isn't.

Let's take the following sequence of events

1. A work package build with a context C that attempts to load storage with preimage P
2. When executed against the state of C, this query returns `None`.
3. In C+1, the value is posted to the storage pallet.
4. When executed against the state of C+k, this query returns `Some`
5. Consensus fault

Reply:
> Sometimes the preimage is available on-chain and needs to look like it isn't, depending on context

Sure. This context would be known by the host environment, which would change the semantics of the `lookup` function accordingly. In general, `lookup` would only work if the number of references for any given preimage is greater than zero. After the number of references becomes zero, the preimage is still stored on-chain and thus available to the environment which is hosting in-core code, but it appears to the code executing in-core as though it is not available. In the case of a non-standard execution (e.g. in the case of a dispute sometime later), this could be changed to allow `lookup` to work for any preimage which is on-chain regardless of the number of references. This logic is already described later in the RFC under the Storage Pallet API.

Reply:
Someone could, in alternating blocks, increment refs from zero to one and back down again and do this over 24 hours for many different storage items. We'd need this information to adequately emulate the action of `lookup` in any given block over the last 24 hours (which could, we presume, yet be disputed).

This will create a lot of state churn and/or bloat and probably isn't sound.

It's fairly easily avoided by introducing a latching mechanism, making the minimum time for which `lookup`'s result may cycle from `None` to `Some` and back to `None` for any given hash to be 24 hours. So after some data's references drops to zero, there's a minimum 24 hour period where it will still work properly following the last time it dropped to zero.

Reply:
> which would change the semantics of the lookup function accordingly.

This is what I was commenting on and isn't clear from the interface as written in the RFC, which appears to be a host function definition. I don't see any need for the _in-core_ code (`fn refine` or `fn is_authorized`) to be able to choose the context. From the in-core code it should simply be `fn lookup`. It's rather that the executing validator needs to invoke a `fn lookup_actual(context, preimage)` as a host function to the relay chain state and this was unclear from the RFC text.

Reply:
Keeping the state around for 24 hours is not enough, because a dispute may happen on a fork which some validators never see as the longest chain and so they never import the relay chain blocks with the state. However mainting an off-chain look up table should work. If the work package is accumulated, then 2/3 of validators voted for availability and they should have imported the relay parent. That means that if the work package is available then asking a random validator for the preimage should work. 

If we never see the relay parent, then we don't know whether the preimage was registered on chain at that particular block. My suggestion is to add to the work package somewhere a list of lookup hashes and proofs from the context that these were registered on the relay chain. This would be longer than a hash, like witness data in current PoV blocks, but it is definitely still gives a huge saving on lookup data the size of a parachain runtime.

These proofs should be verified, and the lookup data acquired if the validator doesn't have it, before actually executing the VM. Since otherwise without this outside the VM, we'd just have to halt execution on the host call that looks up a missing value, which seems awkward if we have to resume or rerun later.

If the work package was available, we should be able to obtain any missing values and output valid or invalid as required. Because the witness data proves that they were in the relay chains state of a block that 2/3 of validators imported. If it was never available, then this becomes trickier. I think in disputes, we pass work packages around directly between validators, and this can work for unavailable work packages. In that case, we should also not  forward a work package unless we have all the lookup data.

Yes, this doesn't seem so simple.

Reply:
I now think the design above is sensible. One of the points of asynchronous backing was to have validate, now refine, possible for a light client of the relay chain. This has practical implications for builders too, it is easier for them to build a valid block on the relay parent. It might even be useful for hierarchical relay chains. But mostly it is just a useful abstraction that makes coding this securely easier.